AX: Merge AccessibilityMediaObject into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 32088add148b39472e6657c197b184fa43825942
<pre>
AX: Merge AccessibilityMediaObject into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=297337">https://bugs.webkit.org/show_bug.cgi?id=297337</a>
<a href="https://rdar.apple.com/158234555">rdar://158234555</a>

Reviewed by Tyler Wilcock.

Merge all of the logic from AccessibilityMediaObject into
Accessibility[Render/Node]Object as part of our effort to flatten our
class hierarchy.

A new static-only class, AccessibilityMediaHelpers, was added to
encapsulate the media element logic.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
* Source/WebCore/accessibility/AccessibilityMediaHelpers.cpp: Added.
(WebCore::AccessibilityMediaHelpers::enterFullscreen):
(WebCore::AccessibilityMediaHelpers::toggleMute):
(WebCore::AccessibilityMediaHelpers::interactiveVideoDuration):
(WebCore::AccessibilityMediaHelpers::isPlaying):
(WebCore::AccessibilityMediaHelpers::isAutoplayEnabled):
(WebCore::AccessibilityMediaHelpers::isMuted):
(WebCore::AccessibilityMediaHelpers::press):
(WebCore::AccessibilityMediaHelpers::increment):
(WebCore::AccessibilityMediaHelpers::decrement):
(WebCore::AccessibilityMediaHelpers::mediaSeek):
* Source/WebCore/accessibility/AccessibilityMediaHelpers.h: Renamed from Source/WebCore/accessibility/AccessibilityMediaObject.h.
* Source/WebCore/accessibility/AccessibilityMediaObject.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::mediaElement const):
(WebCore::AccessibilityNodeObject::videoElement const):
(WebCore::AccessibilityNodeObject::increment):
(WebCore::AccessibilityNodeObject::decrement):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::press):
(WebCore::AccessibilityRenderObject::interactiveVideoDuration const):
(WebCore::AccessibilityRenderObject::toggleMute):
(WebCore::AccessibilityRenderObject::isPlaying const):
(WebCore::AccessibilityRenderObject::isMuted const):
(WebCore::AccessibilityRenderObject::isAutoplayEnabled const):
(WebCore::AccessibilityRenderObject::enterFullscreen const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityIsWebInteractiveVideo]):
(-[WebAccessibilityObjectWrapper interactiveVideoDescription]):
(-[WebAccessibilityObjectWrapper accessibilityIsMediaPlaying]):
(-[WebAccessibilityObjectWrapper accessibilityIsMediaMuted]):
(-[WebAccessibilityObjectWrapper accessibilityToggleMuteForMedia]):
(-[WebAccessibilityObjectWrapper accessibilityVideoEnterFullscreen]):

Canonical link: <a href="https://commits.webkit.org/298682@main">https://commits.webkit.org/298682@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa318e585c76099dcb8c8511ef27cc0abaf5c135

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122338 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66841 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3031d8a5-fcd5-43cc-9115-3dc5eadc3991) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88366 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b2c496d9-d995-4ff8-abc4-550694b8b601) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119230 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104333 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68748 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28313 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22439 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66019 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98607 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125487 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43175 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32426 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97042 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43540 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96831 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24645 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42128 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20017 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39122 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43062 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48654 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42529 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45864 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44233 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -803,7 +802,7 @@ Ref<AccessibilityRenderObject> AXObjectCache::createObjectFromRenderer(RenderObj
 #if PLATFORM(IOS_FAMILY)
         if (is<HTMLMediaElement>(*element) && hasRole(*element, nullAtom()))
-            return AccessibilityMediaObject::create(AXID::generate(), renderer, *this);
+            return AccessibilityRenderObject::create(AXID::generate(), renderer, *this);
```
What would happen if we deleted lines 803 - 806 entirely? Would it create an AccessibilityRenderObject anyway at the bottom of this function because every other test failed?

hoffmanjoshua replied:
This seems to work! We should probably do this more, since things are more dynamic.

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1761,6 +1767,41 @@ bool AccessibilityRenderObject::setValue(const String& string)
+{
+#if PLATFORM(IOS_FAMILY)
+    if (mediaElement()) {
+        mediaSeek(AXSeekDirection::Forward);
```
Maybe this code should go in AccessibilityNodeObject, with the other logic? We may eventually merge those, and it doesn't make sense to have half of the logic in AccessibilityRenderObject and half in AccessibilityNodeObject when the actual object is always going to be the same one now.

Putting this inside a UserGestureIndicator scope is probably a good idea, too.

hoffmanjoshua replied:
Moved this and `decrement` to `AccessibilityNodeObject`, which will let them automatically take advantage of that `UserGestureIndicator`.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2930,4 +2971,70 @@ bool AccessibilityRenderObject::isIgnoredElementWithinMathTree() const
+        return;
+
+    // Step 5% each time.
+    const double seekStep = .05;
```
Could be `constexpr` I think?

hoffmanjoshua replied:
Updated in latest commit

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2930,4 +2971,70 @@ bool AccessibilityRenderObject::isIgnoredElementWithinMathTree() const
+    if (RefPtr element = dynamicDowncast<HTMLVideoElement>(node()))
+        element->enterFullscreen();
+}
+#endif
```
#endif // PLATFORM(IOS_FAMILY)

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2930,4 +2971,70 @@ bool AccessibilityRenderObject::isIgnoredElementWithinMathTree() const
+    return element ? localizedMediaTimeDescription(element->duration()) : String();
+}
+
+void AccessibilityRenderObject::mediaSeek(AXSeekDirection direction)
```
What do you think about putting these in a helper class rather than making AccessibilityRenderObject even larger?

I can think of a few patterns that could work.

It could be a RAII-style class like:

```
AccessibilityMediaElement axMediaElement(mediaElement());
axMediaElement.mediaSeek(AXSeekDirection::Forward);
```

Or it could just be some static functions that are encapsulated, like:

```
AccessibilityMediaElement::mediaSeek(mediaElement(), AXSeekDirection::Forward);
```

twilco replied:
I'd personally prefer the latter wherever applicable — nicer to not have to create an instance of a class just to use some predefined logic (but maybe I'm not totally understanding the benefits of the first approach)

minorninth replied:
It may be that in this particular case it'd be too trivial of an amount of code to split out, if most of the added code is called directly by the wrapper as opposed to internal helpers. So totally fine if you don't take this suggestion.

hoffmanjoshua replied:
New class added for this: `AccessibilityMediaHelpers`!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2930,4 +2971,70 @@ bool AccessibilityRenderObject::isIgnoredElementWithinMathTree() const
 #endif
 
+#if PLATFORM(IOS_FAMILY)
+#pragma mark Media Element Handlers
```
Won't any method we add to the end of the file be added to this `Media Element Handlers` section, even if it's totally unrelated to media elements?

hoffmanjoshua replied:
WebKit tends to use this pattern elsewhere, and authors are responsible for adding new flags. But, I added an empty mark which will clear this after the block, so any method added after will be omitted.

twilco commented on Source/WebCore/accessibility/AccessibilityMediaHelpers.h:
```diff
@@ -30,44 +30,30 @@
-    bool isMuted() const;
+    static bool press(HTMLMediaElement*);
+    static void increment(HTMLMediaElement*);
+    static void decrement(HTMLMediaElement*);
```
All of the methods in this class take pointers. Can they take references? Seems like some contexts have already null-checked the value? For functions that do need to pass a pointer, maybe we can add a pointer overload that does a null-check and and then calls the reference version of the function?

hoffmanjoshua replied:
Updated these three to take references

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -706,50 +705,50 @@ - (BOOL)accessibilityIsWebInteractiveVideo
     // Convey the video object as interactive if auto-play is not enabled.
-    auto* mediaObject = dynamicDowncast<AccessibilityMediaObject>(self.axBackingObject);
-    return mediaObject && !mediaObject->isAutoplayEnabled();
+    RefPtr renderObject = dynamicDowncast<AccessibilityRenderObject>(self.axBackingObject)
```
`renderObject` is maybe a confusing name considering the ubiquity of `RenderObject`. Maybe `axRenderObject` or even just `object`?

hoffmanjoshua replied:
fixed both here and in the method below!

twilco commented on Source/WebCore/accessibility/AccessibilityMediaHelpers.cpp:
```diff
@@ -0,0 +1,112 @@
+    double time = direction == AXSeekDirection::Forward ? std::min(current + timeDelta, duration) : std::max(current - timeDelta, 0.0);
+    mediaElement.setCurrentTime(time);
+}
+
```
Extra newline here I think

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -706,50 +705,50 @@ - (BOOL)accessibilityIsWebInteractiveVideo
 
-    auto* mediaObject = dynamicDowncast<AccessibilityMediaObject>(self.axBackingObject);
-    return mediaObject && mediaObject->isPlaying();
+    RefPtr renderObject = dynamicDowncast<AccessibilityRenderObject>(self.axBackingObject);
```
Couple more `renderObject`s in this file 🙂

twilco reviewed (approved):
🚀

hoffmanjoshua committed:
AX: Merge AccessibilityMediaObject into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=297337
rdar://158234555

Reviewed by Tyler Wilcock.

Merge all of the logic from AccessibilityMediaObject into
Accessibility[Render/Node]Object as part of our effort to flatten our
class hierarchy.

A new static-only class, AccessibilityMediaHelpers, was added to
encapsulate the media element logic.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
* Source/WebCore/accessibility/AccessibilityMediaHelpers.cpp: Added.
(WebCore::AccessibilityMediaHelpers::enterFullscreen):
(WebCore::AccessibilityMediaHelpers::toggleMute):
(WebCore::AccessibilityMediaHelpers::interactiveVideoDuration):
(WebCore::AccessibilityMediaHelpers::isPlaying):
(WebCore::AccessibilityMediaHelpers::isAutoplayEnabled):
(WebCore::AccessibilityMediaHelpers::isMuted):
(WebCore::AccessibilityMediaHelpers::press):
(WebCore::AccessibilityMediaHelpers::increment):
(WebCore::AccessibilityMediaHelpers::decrement):
(WebCore::AccessibilityMediaHelpers::mediaSeek):
* Source/WebCore/accessibility/AccessibilityMediaHelpers.h: Renamed from Source/WebCore/accessibility/AccessibilityMediaObject.h.
* Source/WebCore/accessibility/AccessibilityMediaObject.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::mediaElement const):
(WebCore::AccessibilityNodeObject::videoElement const):
(WebCore::AccessibilityNodeObject::increment):
(WebCore::AccessibilityNodeObject::decrement):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::press):
(WebCore::AccessibilityRenderObject::interactiveVideoDuration const):
(WebCore::AccessibilityRenderObject::toggleMute):
(WebCore::AccessibilityRenderObject::isPlaying const):
(WebCore::AccessibilityRenderObject::isMuted const):
(WebCore::AccessibilityRenderObject::isAutoplayEnabled const):
(WebCore::AccessibilityRenderObject::enterFullscreen const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityIsWebInteractiveVideo]):
(-[WebAccessibilityObjectWrapper interactiveVideoDescription]):
(-[WebAccessibilityObjectWrapper accessibilityIsMediaPlaying]):
(-[WebAccessibilityObjectWrapper accessibilityIsMediaMuted]):
(-[WebAccessibilityObjectWrapper accessibilityToggleMuteForMedia]):
(-[WebAccessibilityObjectWrapper accessibilityVideoEnterFullscreen]):

Canonical link: https://commits.webkit.org/298682@main
